src/mscorlib/shared/System/IO/Path.cs

```diff
@@ -383,11 +383,68 @@ public static string Combine(params string[] paths)
             return StringBuilderCache.GetStringAndRelease(finalPath);
         }
 
-        /// <summary>
-        /// Combines two paths. Does no validation of paths, only concatenates the paths
-        /// and places a directory separator between them if needed.
-        /// </summary>
-        private static string CombineNoChecks(ReadOnlySpan<char> first, ReadOnlySpan<char> second)
+        // Unlike Combine(), Join() methods do not consider rooting. They simply combine paths, ensuring that there
+        // is a directory separator between them.
+
+        public static string Join(ReadOnlySpan<char> path1, ReadOnlySpan<char> path2)
+        {
+            if (path1.Length == 0)
+                return path2.Length == 0 ? string.Empty : new string(path2);
+            if (path2.Length == 0)
+                return new string(path1);
+
+            return JoinInternal(path1, path2);
+        }
+
+        public static string Join(ReadOnlySpan<char> path1, ReadOnlySpan<char> path2, ReadOnlySpan<char> path3)
+        {
+            if (path1.Length == 0)
+                return Join(path2, path3);
+
+            if (path2.Length == 0)
+                return Join(path1, path3);
+
+            if (path3.Length == 0)
+                return Join(path1, path2);
+
+            return JoinInternal(path1, path2, path3);
+        }
+
+        public static bool TryJoin(Span<char> destination, out int charsWritten, ReadOnlySpan<char> path1, ReadOnlySpan<char> path2)
```

Comment:
How about something like this to avoid the extra path1/path2.Length == 0 checks?

In the worst case, there are 3 extra checks in the current impl.

```C#
        public static bool TryJoin(Span<char> destination, out int charsWritten, ReadOnlySpan<char> path1, ReadOnlySpan<char> path2)
        {
            charsWritten = 0;
            if (path1.IsEmpty)
            {
                if (path2.IsEmpty)
                    return true;
                if (destination.Length < path2.Length)
                    return false;

                path2.CopyTo(destination);
                charsWritten = path2.Length;
                return true;
            }
            else if (path2.IsEmpty)
            {
                if (destination.Length < path1.Length)
                    return false;

                path1.CopyTo(destination);
                charsWritten = path1.Length;
                return true;
            }

            Debug.Assert(!path1.IsEmpty && !path2.IsEmpty);

            int charsNeeded = path1.Length + path2.Length;
            if (!(IsDirectorySeparator(path1[path1.Length - 1]) || IsDirectorySeparator(path2[0])))
            {
                charsNeeded++;

                if (destination.Length < charsNeeded)
                    return false;
                path1.CopyTo(destination);
                destination[path1.Length] = DirectorySeparatorChar;
                path2.CopyTo(destination.Slice(path1.Length + 1));
            }
            else
            {
                if (destination.Length < charsNeeded)
                    return false;
                path1.CopyTo(destination);
                path2.CopyTo(destination.Slice(path1.Length));
            }

            charsWritten = charsNeeded;
            return true;
        }
```

Reply:
Because it is a little more bulky & complicated. I can't imagine optimizing checking length being worth it.

Reply:
Good point. What about the needsSeparator branching? Can we just have a single if/else (which results in some code duplication).

Reply:
It does seem to make a big enough difference:
![image](https://user-images.githubusercontent.com/6527137/36758371-15ccde84-1bc9-11e8-8105-78b71ed6379c.png)

How large do we expect the path length to be on average? The smaller the path length, the bigger the difference.

Reply:
There will be all sorts. The first path will typically be longer than the second. I don't want to add the complexity here unless I can see a real impact in a higher level scenario. Directory.GetFiles() over a large set, for example. I'll be switching to use these methods (as opposed to our current helpers) as soon as we drop an updated CLR and we can revisit.